### PR TITLE
Update Kotlin to version 2.1.0

### PR DESCRIPTION
**Pull request by andreas-repp**
This PR updates Kotlin to the [newly released version 2.1.0](https://blog.jetbrains.com/kotlin/2024/11/kotlin-2-1-0-released/)

### CODE REVIEW

**Comment by gsmet** [comment]
Thanks for taking care of this!

**Comment by andreas-eberle** [comment]
Thanks for the reminder. I forgot I didn't push my last commit and undraft this. It should be updated and hopefully the CI passes now.

**Comment by geoand** [comment]
Thanks!

**Review by geoand — APPROVED** [review]

**Comment by geoand** [comment]
CI didn't really run on this which is really weird

**Review by gsmet — CHANGES_REQUESTED** [review]
Thanks for this effort, greatly appreciated.

I have a question though and I would like to get it sorted out before we merge. See inline.

File: integration-tests/resteasy-reactive-kotlin/standard/src/main/kotlin/io/quarkus/it/resteasy/reactive/kotlin/CountryNamePayloadTransformer.kt
Comment (andreas-eberle) [reply integration-tests/resteasy-reactive-kotlin/standard/src/main/kotlin/io/quarkus/it/resteasy/reactive/kotlin/CountryNamePayloadTransformer.kt:15]: Yes, this is actually required. The `toLowerCase` method has been deprecated with a warning for a while but now they removed it or changed it to an error (I don't remember exactly any more, but the project didn't compile any longer). So this is a needed fix and `lowercase()` is the intended replacement.

**Comment by geoand** [comment]
Seems like `quarkus-integration-test-kotlin` is failing because of:

```
2024-12-04T07:18:19.3974147Z Caused by: io.quarkus.builder.BuildException: Build failure: Build failed due to errors
2024-12-04T07:18:19.3977020Z 	[error]: Build step io.quarkus.smallrye.graphql.deployment.SmallRyeGraphQLProcessor#buildExecutionService threw an exception: java.lang.IllegalArgumentException: Provided Metadata instance has version 2.1.0, while maximum supported version is 2.0.0. To support newer versions, update the kotlinx-metadata-jvm library.
2024-12-04T07:18:19.3979808Z 	at kotlinx.metadata.jvm.KotlinClassMetadata$Companion.throwIfNotCompatible$kotlinx_metadata_jvm(KotlinClassMetadata.kt:609)
2024-12-04T07:18:19.3981342Z 	at kotlinx.metadata.jvm.KotlinClassMetadata$Companion.checkMetadataVersionForRead(KotlinClassMetadata.kt:600)
2024-12-04T07:18:19.3982604Z 	at kotlinx.metadata.jvm.KotlinClassMetadata$Companion.read(KotlinClassMetadata.kt:579)
2024-12-04T07:18:19.3983607Z 	at kotlinx.metadata.jvm.KotlinClassMetadata.read(KotlinClassMetadata.kt)
2024-12-04T07:18:19.3984939Z 	at io.smallrye.graphql.schema.creator.OperationCreator.toKotlinClassMetadata(OperationCreator.java:264)
2024-12-04T07:18:19.3986391Z 	at io.smallrye.graphql.schema.creator.OperationCreator.checkWrappedTypeKotlinNullability(OperationCreator.java:126)
2024-12-04T07:18:19.3987866Z 	at io.smallrye.graphql.schema.creator.OperationCreator.createOperation(OperationCreator.java:112)
2024-12-04T07:18:19.3989154Z 	at io.smallrye.graphql.schema.SchemaBuilder.addOperations(SchemaBuilder.java:448)
2024-12-04T07:18:19.3990221Z 	at io.smallrye.graphql.schema.SchemaBuilder.lambda$generateSchema$2(SchemaBuilder.java:159)
2024-12-04T07:18:19.3991175Z 	at java.base/java.util.Optional.ifPresentOrElse(Optional.java:198)
2024-12-04T07:18:19.3992078Z 	at io.smallrye.graphql.schema.SchemaBuilder.generateSchema(SchemaBuilder.java:157)
2024-12-04T07:18:19.3993044Z 	at io.smallrye.graphql.schema.SchemaBuilder.build(SchemaBuilder.java:109)
2024-12-04T07:18:19.3994359Z 	at io.quarkus.smallrye.graphql.deployment.SmallRyeGraphQLProcessor.buildExecutionService(SmallRyeGraphQLProcessor.java:363)
2024-12-04T07:18:19.3995779Z 	at java.base/java.lang.invoke.MethodHandle.invokeWithArguments(MethodHandle.java:732)
2024-12-04T07:18:19.3996623Z 	at io.quarkus.deployment.ExtensionLoader$3.execute(ExtensionLoader.java:856)
2024-12-04T07:18:19.3997400Z 	at io.quarkus.builder.BuildContext.run(BuildContext.java:256)
2024-12-04T07:18:19.3998160Z 	at org.jboss.threads.ContextHandler$1.runWith(ContextHandler.java:18)
2024-12-04T07:18:19.3999290Z 	at org.jboss.threads.EnhancedQueueExecutor$Task.doRunWith(EnhancedQueueExecutor.java:2675)
2024-12-04T07:18:19.4000364Z 	at org.jboss.threads.EnhancedQueueExecutor$Task.run(EnhancedQueueExecutor.java:2654)
2024-12-04T07:18:19.4001695Z 	at org.jboss.threads.EnhancedQueueExecutor.runThreadBody(EnhancedQueueExecutor.java:1627)
2024-12-04T07:18:19.4002858Z 	at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.run(EnhancedQueueExecutor.java:1594)
2024-12-04T07:18:19.4003716Z 	at java.base/java.lang.Thread.run(Thread.java:840)
2024-12-04T07:18:19.4004321Z 	at org.jboss.threads.JBossThread.run(JBossThread.java:499)
```

**Comment by jmartisk** [comment]
Right, I'll try to upgrade `kotlinx.metadata.jvm` in smallrye-graphql, so this will have to wait until a smallrye-graphql upgrade is merged..

**Comment by geoand** [comment]
Thanks!

File: integration-tests/resteasy-reactive-kotlin/standard/src/main/kotlin/io/quarkus/it/resteasy/reactive/kotlin/CountryNamePayloadTransformer.kt
Comment (gsmet) [reply integration-tests/resteasy-reactive-kotlin/standard/src/main/kotlin/io/quarkus/it/resteasy/reactive/kotlin/CountryNamePayloadTransformer.kt:15]: I would expect `uppercase()` to be the replacement for `toUpperCase()` :).

**Comment by andreas-eberle** [comment]
Thanks! Sorry I wasn't able too look for the errors in time. I'm unfortunately a bit swamped right now.

**Comment by gsmet** [comment]
Waiting for https://github.com/smallrye/smallrye-graphql/pull/2234 to land in Quarkus.

**Comment by gsmet** [comment]
@andreas-eberle probably a good idea to include these two changes in this PR:
- https://github.com/quarkusio/quarkus/pull/45215/files
- https://github.com/quarkusio/quarkus/pull/45213/files

A new version of SmallRye GraphQL was released so hopefully we will be able to make progress on this one soon!

**Comment by gsmet** [comment]
Yeah so the GraphQL one won't work as the update breaks things. We will have to wait for the SmallRye GraphQL team to push a proper update.

**Comment by andreas-eberle** [comment]
Thanks for the info @gsmet. Then I'll rebase the PR as soon as the graphql update is properly integrated.

**Comment by jmartisk** [comment]
@andreas-eberle I've submitted https://github.com/quarkusio/quarkus/pull/45228 that also incorporates your commit, I think we need to apply both upgrades at the same time. Please check it

**Comment by andreas-eberle** [comment]
Thanks @jmartisk. Looks good for me from the kotlin update side. Let's see what CI says about your PR. I will close this one.

**Comment by geoand** [comment]
Closing in favor of #45228

**Pull request closed without merging**
